[test] Add OpenTitan target to riscv-compliance testing

tjaychen opened the pull request:
Signed-off-by: Timothy Chen <timothytim@google.com>

imphil commented on riscv-target/opentitan/device/rv32imc/Makefile.include:
```diff
@@ -0,0 +1,76 @@
+# After the libraries are built, the necessary collateral (vmem for verilator, bin
+# for fpga) are created
+COMPILE_TARGET=\
+        make -C $$(OTSW) SW_DIR=boot_rom $(MAKEARG) clean all; \
```
Please fix the alignment here.

tjaychen replied:
fixed

imphil commented on riscv-target/opentitan/device/rv32imc/Makefile.include:
```diff
@@ -0,0 +1,76 @@
+RV_TOOLS    ?= /tools/riscv/bin
+FPGA_UART   ?=
+TARGET_SIM  ?= $(OTROOT)/build/lowrisc_systems_top_earlgrey_verilator_0.1/sim-verilator/Vtop_earlgrey_verilator
+TARGET      ?= fpga
```
Can you use variable names which make it very clear which ones are opentitan-specific and which ones are the ones commonly used in the riscv-compliance test suite? E.g. prefix all opentitan variables with `OT_` or `OPENTITAN_`.

tjaychen replied:
fixed, at least IMO. Let me know if you prefer other names

gkelly commented on riscv-target/opentitan/README.md:
```diff
@@ -0,0 +1,135 @@
+
+By default, the test assumes there exists a valid Verilator build at `${REPO_TOP}/build/lowrisc_systems_top_earlgrey_verilator_0.1/sim-verilator/Vtop_earlgrey_verilator`.
+If your Verilator build is at a different location, please set that as well when running with Verilator.
+For instructions on how to create a Verilator build, please refer to the [Verilator guide]({{< relref "doc/ug/getting_started_verilator" >}}).
```
You'll have to link to the page on docs.opentitan.org https://docs.opentitan.org/doc/ug/getting_started_verilator/

tjaychen replied:
yep will do.

gkelly commented on riscv-target/opentitan/README.md:
````diff
@@ -0,0 +1,135 @@
+```
+
+When running against FPGA, the test assumes the FPGA is already programmed and ready to go.
+To find out how to properly build and flash FPGA, please refer to the [FPGA manual]({{< relref "doc/rm/ref_manual_fpga" >}})
````
https://docs.opentitan.org/doc/rm/ref_manual_fpga/

tjaychen replied:
i'm taking out all the links here and referring directly to quick_start as @imphil suggests.

gkelly commented on riscv-target/opentitan/device/rv32imc/link.ld:
```diff
@@ -0,0 +1,55 @@
+ENTRY(_start)
+
+/* required to correctly link newlib */
+GROUP( -lc -lgloss -lgcc -lsupc++ )
```
Can you format these parenthetical arguments cuddled to the braces `("riscv")`, `(-lc`, etc

tjaychen replied:
fixed

gkelly commented on riscv-target/opentitan/device/rv32imc/link.ld:
```diff
@@ -0,0 +1,55 @@
+MEMORY
+{
+    flash (rx)     : ORIGIN = 0x20000000, LENGTH = 0x100000
+    ram (wx)      : ORIGIN = 0x10000000, LENGTH = 0x10000
```
nit: Line up the `:`s.

tjaychen replied:
fixed

gkelly commented on riscv-target/opentitan/device/rv32imc/link.ld:
```diff
@@ -0,0 +1,55 @@
+  .text.trap : {
+        . = ALIGN(0x100);
+        *(.text.trap)
+  } > flash
```
Fix indentation.

tjaychen replied:
i'm not exactly sure what's expected here, so i just put 2 spaces for all of them.

gkelly commented on riscv-target/opentitan/device/rv32imc/wrap.c:
```diff
@@ -0,0 +1,38 @@
+
+void dump_signature(void) {
+  extern uint32_t begin_signature[];
+  extern uint32_t end_signature[];
```
It's strange to declare externs inside a function like this, can you move them out of the function scope.

tjaychen replied:
fixed

gkelly commented on riscv-test-suite/rv32i/Makefrag:
```diff
@@ -13,16 +13,16 @@
 # LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 # ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
-# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF 
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
```
No need to change the formatting here.

tjaychen replied:
i didnt, not intentionally.  It was just trailing whitespace stripping from my editor.
I thought it would be a bit of overkill to go back and add it back...

gkelly commented on riscv-test-suite/rv32i/Makefrag:
```diff
@@ -80,8 +77,6 @@ rv32i_sc_tests = \
     I-CSRRC-01 \
     I-CSRRCI-01 \
-    I-ECALL-01 \
-    I-EBREAK-01 \
```
Why are these removed?  Are they just tests that OpenTitan currently fails?

tjaychen replied:
I should have clarified. These are tests that currently fail on ibex. 
so they are also removed for OpenTitan.  I'll add a note in the readme.

gkelly replied:
Can you move them to the bottom and comment them out?  Just to make it easier to re-enable them in the future?

tjaychen replied:
moved to the bottom. please take a took.  Also added to README.

tjaychen commented:
> Thanks for working on this @tjaychen. The wrapping and dumping of the signature looks very nice. It would be nice to get the style of some of the files more aligned to the riscv-compliance upstream, since we ultimately want to get this code merged there.
> 
> What's missing currently is a timeout mechanism, as we have it in Ibex. What's your plan on that?
> 
> I'm a bit worried about the way these tests are called now, but that's non-blocking for now.
> Most of my concern comes from the fact that we require people to have an opentitan source tree with a fragile set of binaries and other things in un(der)documented places around the source tree.
> Working on that is beyond the scope of this PR and tracked in [lowRISC/opentitan#650](https://github.com/lowRISC/opentitan/issues/650). Once we got that fixed we can go back to this code.

yeah i agree about the binaries part.  Hopefully that's something that will be cleaned up over time.  I'm also hoping to eventually massage riscv compliance into the DV construct, so that will bring it more inline with how things should officially behave.

Regarding timeout, one of the benefit of switching to @mcy pytest implementation is that there are native timeouts (1 minute).  So when I previously still had the failing tests, they would timeout and continue.  This isn't the perfect solution, since it'll vary depending in design complexity, but i think it suits our needs for now.

tjaychen commented on riscv-target/opentitan/README.md:
```diff
@@ -0,0 +1,135 @@
+---
+title: "OpenTitan RISC-V Compliance Testing"
+---
```
yeah good point. I realized that yesterday but forgot to change it.

tjaychen commented on riscv-target/opentitan/README.md:
```diff
@@ -0,0 +1,135 @@
+
+# Overview
+The RISC-V compliance test can be run on either FPGA or Verilator.
+To run on Verilator, set the variables below
```
Would it be okay if I just linked to the main website?

tjaychen commented on riscv-target/opentitan/README.md:
````diff
@@ -0,0 +1,135 @@
+```console
+$ export RISCV_TARGET=opentitan
+$ export RISCV_DEVICE=rv32imc
+$ export TARGET=verilator
````
as a separate note..i'm beginning to think OPENTITAN might be too broad.. especially if long term OPENTITAN supports multiple tops.  I think however that's probably a better discussion for another time / thread.

tjaychen commented on riscv-target/opentitan/README.md:
```diff
@@ -0,0 +1,135 @@
+
+By default, the test assumes there exists a valid Verilator build at `${REPO_TOP}/build/lowrisc_systems_top_earlgrey_verilator_0.1/sim-verilator/Vtop_earlgrey_verilator`.
+If your Verilator build is at a different location, please set that as well when running with Verilator.
+For instructions on how to create a Verilator build, please refer to the [Verilator guide]({{< relref "doc/ug/getting_started_verilator" >}}).
```
can i just reference the quick-start? I really want to avoid duplicating step by step instructions everywhere...because when anything changes, there's such a large fanout.

tjaychen commented on riscv-target/opentitan/README.md:
```diff
@@ -0,0 +1,135 @@
+$ cd $REPO_TOP
+$ make -C sw/vendor/riscv_compliance RISCV_ISA=rv32i \
+  && make -C sw/vendor/riscv_compliance RISCV_ISA=rv32im \
+  && make -C sw/vendor/riscv_compliance RISCV_ISA=rv32imc
```
that's actually what I had originally, but a separate review asked me to change to this form.  Thinking about it more though, I think what you're saying does make more sense from the perspective of riscv-compliance

tjaychen replied:
fixed

tjaychen commented on riscv-target/opentitan/README.md:
```diff
@@ -0,0 +1,135 @@
+Check         I-XOR-01 ... OK
+Check        I-XORI-01 ... OK
+--------------------------------
+OK: 55/55
```
sounds good.

tjaychen commented on riscv-target/opentitan/compliance_test.h:
```diff
@@ -0,0 +1,39 @@
+#ifndef _COMPLIANCE_TEST_H
+#define _COMPLIANCE_TEST_H
+
+#include "sw/vendor/riscv_compliance/riscv-test-env/p/riscv_test.h"
```
updated

tjaychen commented on riscv-target/opentitan/device/rv32imc/Makefile.include:
```diff
@@ -0,0 +1,76 @@
+## Copyright lowRISC contributors.
+## Licensed under the Apache License, Version 2.0, see LICENSE for details.
+## SPDX-License-Identifier: Apache-2.0
```
fixed

tjaychen commented on riscv-target/opentitan/device/rv32imc/Makefile.include:
```diff
@@ -0,0 +1,76 @@
+## SPDX-License-Identifier: Apache-2.0
+
+
+OT           = $(ROOTDIR)/riscv-target/$(RISCV_TARGET)/device/rv32imc
```
fixed

tjaychen commented on riscv-target/opentitan/device/rv32imc/platform.yaml:
```diff
@@ -0,0 +1,10 @@
+## Copyright lowRISC contributors.
+## Licensed under the Apache License, Version 2.0, see LICENSE for details.
+## SPDX-License-Identifier: Apache-2.0
```
fixed

gkelly commented:
Thanks so much for this change!

tjaychen commented:
@imphil 
can you have more look to see if I've addressed your concerns? Once this is in, I can work on getting CI merged, and probably work towards an FPGA solution as well.
